File: src/dmsg_types/tests/support/commerce.rs
#![allow(dead_code)]
#[path = "integration.rs"]
pub mod base;
use base::*;
use dmsg_protocol::commerce_v2::{
    checkout_id, checkout_quote, checkout_quote_hash, panda_application_hash,
};
use dmsg_protocol::integration::quote_panda;
use dmsg_types::integration::{
    ApprovalPurpose, ProductDecision, SettlementSource, CASH_ACTIVATION_MS,
};
use dmsg_types::integration_billing::{
    CheckoutQuote, OpenCheckout, ProductAuthorizationRequest, SettlementAsset, SettlementAssetKind,
    SubscriptionSource,
};
use dmsg_types::integration_membership::{PandaApplicationTerms, PandaClaimRequest};
use dmsg_types::{AccountId, Environment, Hash, MINUTE};
pub fn asset() -> SettlementAsset {
    SettlementAsset {
        version: 2,
        policy_version: 1,
        environment: Environment::Local,
        ledger: principal(6),
        asset: SettlementAssetKind::CkUsdc,
        decimals: 6,
        price_usd_micros: 1_000_000,
        price_observed_at_ms: NOW,
        price_valid_until_ms: NOW + 30 * MINUTE,
        network_fee_atomic: 10,
        max_network_fee_atomic: 20,
        enabled: true,
    }
}
pub fn checkout() -> CheckoutQuote {
    checkout_quote(
        principal(8),
        offer(),
        &app(),
        &product(),
        asset(),
        principal(9).into(),
        NOW,
    )
    .unwrap()
}
pub fn terms() -> PandaApplicationTerms {
    PandaApplicationTerms {
        home_membership: principal(10),
        user_home: principal(1),
        approving_account: AccountId([1; 12]),
        actor: principal(9),
        sns_governance: principal(11),
        neuron_id: Hash::new([30; 32]),
        offer: offer(),
        quote: quote_panda(&offer(), &app(), &product(), &rate(), NOW).unwrap(),
    }
}
pub fn authorization(cash: bool) -> ProductAuthorizationRequest {
    let mut a = approval();
    a.purpose = if cash {
        ApprovalPurpose::CashCheckout
    } else {
        ApprovalPurpose::PandaSubscription
    };
    a.service = if cash { principal(8) } else { principal(10) };
    a.action_digest = if cash {
        checkout_quote_hash(&checkout())
    } else {
        panda_application_hash(&terms())
    };
    ProductAuthorizationRequest {
        offer: offer(),
        account_approval: a,
        user_home: principal(1),
        approval_id: Hash::new([31; 32]),
        product_approval: None,
    }
}
pub fn open() -> OpenCheckout {
    OpenCheckout {
        quote: checkout(),
        authorization: authorization(true),
    }
}
pub fn claim() -> PandaClaimRequest {
    PandaClaimRequest {
        terms: terms(),
        authorization: authorization(false),
    }
}
pub fn cash_decision() -> ProductDecision {
    let q = checkout();
    ProductDecision {
        version: 2,
        offer: q.offer,
        decision_id: Hash::new([32; 32]),
        source: SettlementSource::Cash {
            order_id: checkout_id(principal(8), &offer()),
            ledger: q.cash.ledger,
            block_index: 0,
            amount_atomic: q.cash.amount_atomic,
        },
        decided_at_ms: NOW,
        apply_by_ms: NOW + CASH_ACTIVATION_MS,
    }
}
pub fn cash_source(d: &ProductDecision) -> SubscriptionSource {
    if let SettlementSource::Cash {
        order_id,
        ledger,
        block_index,
        amount_atomic,
    } = d.source
    {
        SubscriptionSource::Cash {
            order_id,
            ledger,
            block_index,
            amount_atomic,
        }
    } else {
        panic!("cash fixture")
    }
}

File: packages/dmsg-sdk/src/commerce.ts
import { digest, equalBytes, canonical } from "./encoding.ts";
import {
  requireValid,
  validateShape,
  validateRatePolicy,
  validateBillingOffer,
  requiredPandaStake,
  CKUSDC_LEDGER,
  CKUSDT_LEDGER,
} from "./validation.ts";
import { Principal } from "@icp-sdk/core/principal";
import type {
  AppRegistration,
  ProductRegistration,
  CheckoutRequest,
  CheckoutQuote,
  SettlementAsset,
  PandaApplicationTerms,
} from "./contracts.ts";
export const checkoutId = (home: Uint8Array, request: CheckoutRequest) =>
  digest("dmsg/checkout/order/v2", [
    home,
    request.offer.app_id,
    request.offer.operation_id,
  ]);
export const checkoutQuoteHash = (quote: CheckoutQuote) =>
  digest("dmsg/checkout/quote/v2", quote);
export const pandaApplicationHash = (terms: PandaApplicationTerms) =>
  digest("dmsg/panda/application/v2", terms);
export const pandaClaimId = (terms: PandaApplicationTerms) =>
  digest("dmsg/panda/claim/v2", [
    terms.home_membership,
    terms.offer.app_id,
    terms.offer.operation_id,
  ]);
export function validateCheckoutRequest(
  request: CheckoutRequest,
  app: AppRegistration,
  product: ProductRegistration,
  now: bigint,
) {
  validateShape("CheckoutRequest", request);
  validateBillingOffer(request.offer, app, product, now);
  requireValid(
    request.offer.allowed_settlement_methods.includes(request.method),
    "FORBIDDEN",
  );
  if (request.product_approval)
    requireValid(
      request.product_approval.method === request.method &&
        request.product_approval.expires_at_ms > now,
      "POLICY_STALE",
    );
}
export function validateSettlementAsset(asset: SettlementAsset, now: bigint) {
  validateShape("SettlementAsset", asset);
  requireValid(
    asset.version === 2n &&
      asset.policy_version > 0n &&
      asset.decimals === 6n &&
      asset.enabled &&
      asset.price_observed_at_ms <= now &&
      now < asset.price_valid_until_ms &&
      asset.price_valid_until_ms - asset.price_observed_at_ms <= 1_800_000n &&
      asset.price_usd_micros >= 990_000n &&
      asset.price_usd_micros <= 1_010_000n &&
      asset.network_fee_atomic > 0n &&
      asset.network_fee_atomic <= asset.max_network_fee_atomic &&
      asset.max_network_fee_atomic <= 10_000_000n,
    "POLICY_STALE",
  );
  if (asset.environment !== "Local")
    requireValid(
      Principal.fromUint8Array(asset.ledger).toText() ===
        (asset.asset === "CkUsdc" ? CKUSDC_LEDGER : CKUSDT_LEDGER),
      "FORBIDDEN",
    );
}
export async function validateCheckoutQuote(
  quote: CheckoutQuote,
  request: CheckoutRequest,
  app: AppRegistration,
  product: ProductRegistration,
  home: Uint8Array,
  payer: Uint8Array,
  now: bigint,
) {
  validateShape("CheckoutQuote", quote);
  validateCheckoutRequest(request, app, product, now);
  validateSettlementAsset(quote.asset, now);
  const c = quote.cash,
    asset = quote.asset,
    amount =
      (request.offer.amount_usd_micros * 1_000_000n +
        asset.price_usd_micros -
        1n) /
      asset.price_usd_micros;
  requireValid(
    request.method === "Cash" &&
      equalBytes(canonical(request.offer), canonical(quote.offer)) &&
      equalBytes(canonical(quote.product), canonical(product)) &&
      quote.quoted_at_ms <= now &&
      equalBytes(c.ledger, asset.ledger) &&
      product.ledgers.some((p) => equalBytes(p, c.ledger)) &&
      c.amount_atomic === amount &&
      c.fee_reserve_atomic === asset.max_network_fee_atomic * 2n &&
      c.max_network_fee_atomic === asset.max_network_fee_atomic &&
      equalBytes(c.payer.owner, payer) &&
      c.payer.subaccount === null &&
      equalBytes(c.deposit.owner, home) &&
      c.deposit.subaccount !== null &&
      equalBytes(c.deposit.subaccount, await checkoutId(home, request)) &&
      c.funding_deadline_ms ===
        min(quote.quoted_at_ms + 1_800_000n, request.offer.expires_at_ms) &&
      c.activation_deadline_ms ===
        min(quote.quoted_at_ms + 86_400_000n, request.offer.expires_at_ms) &&
      equalBytes(
        c.offer_hash,
        await digest("dmsg/commerce/offer/v2", request.offer),
      ) &&
      equalBytes(
        c.conversion_hash,
        await digest("dmsg/asset-policy/v2", asset),
      ),
    "INTEGRITY_FAILED",
  );
}
export async function validatePandaTerms(
  terms: PandaApplicationTerms,
  request: CheckoutRequest,
  membership: Uint8Array,
  home: Uint8Array,
  actor: Uint8Array,
  neuron: Uint8Array,
  now: bigint,
) {
  validateShape("PandaApplicationTerms", terms);
  const q = terms.quote;
  validateRatePolicy(q.policy);
  requireValid(
    q.version === 2n &&
      q.policy.effective_at_ms <= q.quoted_at_ms &&
      q.policy.product_ids.includes(request.offer.product_id) &&
      q.application_deadline_ms ===
        min(q.quoted_at_ms + 86_400_000n, request.offer.expires_at_ms),
    "INTEGRITY_FAILED",
  );
  requireValid(
    request.method === "Panda" &&
      equalBytes(canonical(terms.offer), canonical(request.offer)) &&
      equalBytes(terms.home_membership, membership) &&
      equalBytes(terms.user_home, home) &&
      equalBytes(terms.actor, actor) &&
      equalBytes(terms.neuron_id, neuron) &&
      equalBytes(terms.approving_account, request.approving_account) &&
      q.quoted_at_ms <= now &&
      now < q.application_deadline_ms &&
      q.committed_until_ms === request.offer.expires_at_ms &&
      q.required_stake_e8s ===
        requiredPandaStake(
          request.offer.amount_usd_micros,
          q.policy.r_num,
          q.policy.r_den,
        ) &&
      equalBytes(
        q.offer_hash,
        await digest("dmsg/commerce/offer/v2", request.offer),
      ),
    "INTEGRITY_FAILED",
  );
}
const min = (a: bigint, b: bigint) => (a < b ? a : b);
